### PR TITLE
Overhaul HA NFS guide

### PR DESCRIPTION
**Pull request by tahliar**
### PR creator: Description

The approach detailed in the NFS guide was potentially harmful, as detailed in https://www.suse.com/support/kb/doc/?id=000020396. 

This new version describes a different approach that hopefully avoids the common pitfalls. 

Here is a PDF for anyone who wants to see a rendered version in addition to the diff: 
[article-nfs-storage_en.pdf](https://github.com/SUSE/doc-sleha/files/11539494/article-nfs-storage_en.pdf) **(updated 2023-05-23)**

### PR creator: Are there any relevant issues/feature requests?

* bsc#1201271
* jsc#DOCTEAM-473

### PR creator: Which product versions do the changes apply to?

When opening a PR, check all versions of the documentation that your PR applies to.

- SLE-HA 15
  - [x] 15 next *(current `main`, no backport necessary)*
  - [x] 15 SP4
  - [x] 15 SP3
  - [x] 15 SP2
  - [x] 15 SP1
- SLE-HA 12
  - [x] 12 SP5
  - [x] 12 SP4
  
**Note** that I'll need to change `clone` back to `ms` for versions 12 SP5 and SP4.

### PR reviewer only: Have all backports been applied?

The doc team member merging your PR will take care of backporting to older documents.
When opening a PR, do *not* set the following check box.

- [ ] all necessary backports are done

### CODE REVIEW

File: xml/article_nfs_storage.xml
Comment (tahliar) [inline xml/article_nfs_storage.xml:198]: @zzhou1 I changed the initial parts of the DRBD config, but left the rest the same as it already was (see https://documentation.suse.com/sle-ha/15-SP4/html/SLE-HA-all/article-nfs-storage.html#sec-ha-quick-nfs-drbd-config).

Let me know if I do need to include the extra details from the configs you attached in the bug.

File: xml/article_nfs_storage.xml
Comment (tahliar) [inline xml/article_nfs_storage.xml:385]: Is this section (Adjusting Pacemaker's configuration) still required?

File: xml/article_nfs_storage.xml
Comment (tahliar) [inline xml/article_nfs_storage.xml:625]: It's my understanding from the discussion in the bug that this line covers point 6 from the TID (consistent host name). I may have misunderstood, though, so let me know if we do still need a note about the `/etc/systemd/system/nfs-server.service.d/cluster.conf` config.

File: xml/article_nfs_storage.xml
Comment (tahliar) [inline xml/article_nfs_storage.xml:725]: I've put this note here in Chapter 7 Using the NFS service, as I originally suggested, but maybe it's better in Section 4.3 Creating the file system? Let me know what you think.

I also skimmed through the Admin guide and didn't notice anywhere that seemed like a good place to also add this note, but I'm open to suggestions.

**Review by tahliar — COMMENTED** [review]
Some open items:

File: xml/article_nfs_storage.xml
Comment (zzhou1) [reply xml/article_nfs_storage.xml:198]: Sounds good to me. Instead of copy&paste among DOCs.

File: xml/article_nfs_storage.xml
Comment (zzhou1) [reply xml/article_nfs_storage.xml:385]: Good catch. I have the same comment and do suggest to delete it. 

Loop in @gao-yan to double check his view.

File: xml/article_nfs_storage.xml
Comment (zzhou1) [reply xml/article_nfs_storage.xml:625]: You are right. `nfs_server_scope`  is the implementation of RA nfsserver to cover the same purpose of `/etc/systemd/system/nfs-server.service.d/cluster.conf`.

File: xml/article_nfs_storage.xml
Comment (zzhou1) [reply xml/article_nfs_storage.xml:725]: > I've put this note here in Chapter 7 Using the NFS service, as I originally suggested, but maybe it's better in Section 4.3 Creating the file system? Let me know what you think.
> 

However, I prefer to move this WARNING under "primitive p_nfsserver nfsserver".  Simply because  NFSV4LEASETIME in `/etc/sysconfig/nfs` is directly used by the nfs server, instead of the nfs client.

**Comment by zzhou1** [comment]
Additional comments:

1. "Section 3 Creating LVM devices" 
  20G /var/lib/nfs is too large comparing to 20G exportfs.  Mightbe 20% is a fair estimation, hence 4G in this example.  Well, @Darcy might have better practical suggestion in the fields.

2. `drbdadm new-current-uuid --clear-bitmap nfs-exportfs/1`

  I spent some minutes and realized the above line combines two forms actually,
  and the digital number is the equavalent drbd minor of drbd res name. 

  `drbdadm new-current-uuid --clear-bitmap nfs-exportfs`
  `drbdadm new-current-uuid --clear-bitmap 1`

  Not sure if it's better to reformat a little bit
 `drbdadm new-current-uuid --clear-bitmap { nfs-exportfs | 1 }`

3. "Section 4.3 Creating the file system"
  ... the same for nfs stat
  mkfs.ext4 /dev/drbd2

4. It makes sense to me to delete `meta target-role=Started`, which is the default.

5. For `primitive p_exportfs exportfs``

  It makes sense to me to have the side note underneath that no need to set
  `wait_for_leasetime_on_stop`, let its default as "false", because of `/var/lib/nfs` is HA-safe.

  Since this parameter has created quit some confusing challenges in the field.

6. In the end of "Section 7 Using the NFS service", it might be worthwhile to add further lines to indicate only NFSv3 supports `loopback mount`, and not for NFSv4.x. Also add Darcy's valuable TID:  
  https://www.suse.com/support/kb/doc/?id=000018709

**Comment by dpartrid** [comment]
I reviewed the PDF, but I don't know if this is the latest copy including all previous edits.

I'm giving feedback as the 'NFS expert' who requested the doc changes.  I'm not really an HA expert.  So while I can say that the HA configuration instructions here *look like* they are accomplishing the points that are important to NFS, I myself am not familiar enough with HA syntax (and I'm not testing these steps myself) so my "approval" must be taken with a grain of salt.  I hope someone with deep HA experience is reading and actually testing these steps.

Next, let me admit that I'm a bit obsessive when it comes to things that could possibly be misunderstood.  I see a few things that worry me in the terminology or naming used.  So some of my suggestions might seem rather nit-picky.

1.  The doc uses the name "exportfs" for one of the volumes.  Thus that name, plus other names built upon it, are used very frequently:

i.e.
volume name exportfs
resource name nfs-exportfs
drdb primitive name drdb_nfs-exportfs
fs primitive name fs_nfs-exportfs
	
Those by themselves are not really a concern until you consider the fact that we will also be using an exportfs primitive (which accomplishes the actual export of the file system).  In other words, "exportfs" is a keyword used when defining export primitives (per your procedure #6, steps #1 and #2), which I quote in part:

primitive p_exportfs exportfs \
params directory="/srv/nfs/exportfs" \

It is also a binary command  name which gets used in procedure 6 step 3.

I think the word "exportfs" is being used in too many different places in this doc, to represent different components of the configuration.  I think it would be best to only use it specific to the exportfs primitive (and obviously also for the exportfs -v command), and to change the name of the volume (and therefore all the other names build upon that volume) to something else (maybe "share1").  Otherwise, this might cause administrators confusion about why the name "exportfs" is being suggested or what it really impacts.  It seems to imply a [non-existent] requirement of matching terms between the volume names and certain parts of the exportfs primitive terminology.

I realize that changing this in the doc must be done very carefully, as a simple "replace all" would replace too much.  But I think it would be worth the effort.

2.  There are a few places in the doc where you refer to the ext4 file system as an "NFS file system".  As someone who deals with NFS discussions and terminology on a constant basis, this bothers me.  An ext4 file system is not an NFS file system.  I would say at the most technical level, the term "NFS file system" refers to a NFS client mount of "type nfs" but does not accurately refer to a fs of "type ext4", even if that file system is going to be exported via NFS.  I might call the ext4 file system an "nfs-shared file system" but I would not call it an "nfs file system".

So for example, at:

     4.3 Creating the file system
       After activating the DRBD devices, create the NFS file system on /dev/drbd1 :

The above 2nd line might be better worded as:

        After activating the DRBD devices, create the ext4 file system (which will later be exported via NFS) on /dev/drbd1 :

Another example:

File system resources
These resources manage the NFS file system and file states.

2nd line could instead say:
These resources manage the ext4 file system which will be exported, and the file system where NFS client states will be tracked.

Another:

6.2 Creating NFS file system resources
Next, create cluster resources to manage the NFS file system and file states.

2nd line could instead say:
Next, create cluster resources to manage the files systems for export and for state tracking.


3.  In section 4.3, I see where you instruct the user to create a ext4 file system on drbd1:
mkfs.ext4 /dev/drbd1

but I don't see anywhere that taken the same kind of step for a file system on drbd2 for the state tracking.  I assume this is just an oversight?

Have any other steps needed for the state file system been missed?  I haven't noticed any others, but I'm not sure I know everything that is needed in the DRBD and HA world.

4.  In section 7, for NFS client mounts, it reads:

In its simplest form, the command to mount the NFS export looks like this:
# mount -t nfs 192.168.1.11:/srv/nfs/exportfs /home/exportfs

Technically, "-t nfs" is not needed.  The mount command recognizes NFS syntax without the the -t designation being present.  So it is not really part of the "simplest" syntax.  Side note:  If "-t nfs" were really needed, then your 2nd example of mount syntax would need to show it as well... but it doesn't need to.

Really, the simplest example would be:

# mount 192.168.1.11:/srv/nfs/exportfs /home/exportfs

5.  Continuing further in section #7, the doc reads:

To configure a specific transport protocol ( proto ) and maximum read and write request sizes
( rsize and wsize ), use:
# mount -o rsize=32768,wsize=32768 \
192.168.1.11:/srv/nfs/exportfs /home/exportfs

But your example didn't actually include the "proto" option which you mentioned.  I would also recommend making it clear that these are just optional examples:

To configure other possible mount options, such as a specific transport protocol ( proto ) and maximum read and write request sizes ( rsize and wsize ), use the -o option.  Note that this is a hypothetical example, not a recommendation:

# mount -o proto=tcp,rsize=32768,wsize=32768 \
192.168.1.11:/srv/nfs/exportfs /home/exportfs


6.  Then after the above mount example, it reads:

If you need to be compatible with NFS version 3, include the value vers=3 after the -o option.
For further NFS mount options, refer to the nfs man page.

"need to be compatible" seems like a potentially misleading term.  To the uninformed, it could be read to imply that maybe you could do mount of some higher version, but make sure it is also backward compatable with NFS v3.  We could still tell people how to set a version, but I'd prefer it was without explicitly suggesting just a specific one.  Maybe:

If the NFS client needs to use a certain version of NFS, include a vers=n value in the comma-separated list of options after -o.

But to be clear:  At this point, I'm not trying to encourage nor discourage the use of v3.

As to some of zzhou1's comments:

I don't know how much space /var/lib/nfs might end up needing. I suspect 20 GB is far more than is needed, but I've sent a message to Neil Brown asking his opinion.

I agree that it would be a nice precaution to note under "primitive p_exportfs exportfs ..." that "wait_for_leasetime_on_stop=true" should not be used, even though many cluster guides recommend it.

About loopback mounts:  I wasn't originally thinking to include this, but I must admit it is useful information.  You could note something like the following:

Some clusters are designed to export an NFS share and then NFS mount it again on the same node.  When a node NFS-mounts its own NFS share, this is called a "NFS loopback mount".  These mounts are only supported under NFSv3.  See https://www.suse.com/support/kb/doc/?id=000018709

**Comment by tahliar** [comment]
Thank you Roger and Darcy for all your feedback! I'll work through it and have an update (with new PDF) either today or early next week. 


> Next, let me admit that I'm a bit obsessive when it comes to things that could possibly be misunderstood. I see a few things that worry me in the terminology or naming used. So some of my suggestions might seem rather nit-picky.

Please keep being nit-picky! Your insights are always really valuable.

File: xml/article_nfs_storage.xml
Comment (tahliar) [reply xml/article_nfs_storage.xml:625]: Great, thanks for confirming!

File: xml/article_nfs_storage.xml
Comment (tahliar) [reply xml/article_nfs_storage.xml:725]: That makes sense to me too. @dpartrid what do you think?

**Comment by tahliar** [comment]
I've made a number of fixes, which can be seen on their own in commit 8d09d03c815675eb639eb2802a7afa2d527e3547, or as part of the whole in the diff. I also updated the PDF with these changes. 

I do expect I'll need to make more changes, either from this round or from later feedback from Neil and/or testing. 

Semi-open items from this round: 

- I deleted the chapter `Adjusting Pacemaker's configuration` on agreement with Roger, but can restore it if Yan disagrees and says it's still needed. 
- I moved the `NFSV4LEASETIME` warning to Section `4.3 Creating the file systems` (it could also work under the `nfsserver` step in `Procedure 7: Creating supporting resources for NFS`, but I didn't want to clog it with notes). But I can move it if this still isn't the best place.
- I changed the size of the `state` LVM volume to 4GB, but can change it if Neil gives a different recommendation.

File: xml/article_nfs_storage.xml
Comment (zzhou1) [inline xml/article_nfs_storage.xml:280]: Also the port number should be different than 7790, say use 7791

File: xml/article_nfs_storage.xml
Comment (zzhou1) [inline xml/article_nfs_storage.xml:664]: Move this step to verify the NFS export later, only after `sudo crm configure commit`.

File: xml/article_nfs_storage.xml
Comment (zzhou1) [inline xml/article_nfs_storage.xml:626]: Probably one more prerequisite to info user to run, just in case, 
`zypper in nfs-kernel-server`

**Review by zzhou1 — CHANGES_REQUESTED** [review]

File: xml/article_nfs_storage.xml
Comment (dpartrid) [reply xml/article_nfs_storage.xml]: Yes, it looks good.  (And this was indeed within the PDF I reviewed.)

File: xml/article_nfs_storage.xml
Comment (dpartrid) [reply xml/article_nfs_storage.xml:340]: The /1 and /2 puzzled me as well.  I'm glad you're suggesting the change.

File: xml/article_nfs_storage.xml
Comment (dpartrid) [reply xml/article_nfs_storage.xml:664]: I wondered about that also, this time through.   Good catch.

File: xml/article_nfs_storage.xml
Comment (tahliar) [reply xml/article_nfs_storage.xml:280]: I guess this also means you need to open port 7791 as well in step 1 of `Activating the DRBD devices`?

Comment (zzhou1) [reply xml/article_nfs_storage.xml:305]: Now pdf is correct. port 7790 is for nfs-share, and port 7791 is for nfs-state.

File: xml/article_nfs_storage.xml
Comment (zzhou1) [inline xml/article_nfs_storage.xml:289]: Hold on, 

The resource-and-stonith with crm-fence-peer.sh doesn't work well for multiple DRBD resources. With that, I incline to combine both resources into one, while still leave two volumes one for nfs-share, one for nfs-state.

That says, the design for multiple DRBD resources will need have more advanced consideration and trade-off, and might not fit this preliminary document.

Bottom line, the whole setup need more rigorous testing.

Comment (dpartrid) [reply xml/article_nfs_storage.xml:289]: I may be misunderstanding this comment.  We often have people using more than 2 of more file systems, shared via NFS, in one cluster.  If certain things don't work well for multiple DRDB resources, then that could be a problem as well, even without considering the "state" resource.  I'd like our steps to be scalable.  With multiple files systems shared via NFS, is there a way to combine those all into one resource as well, for purposes of this concern?

File: xml/article_nfs_storage.xml
Comment (dpartrid) [reply xml/article_nfs_storage.xml:305]: Thanks.  I wasn't fully following this concern before, but seeing the new version, I get it.

**Comment by dpartrid** [comment]
I've heard back from Neil about the size of /var/lib/nfs.  It's kind of a mixed message. I will quote him below but insert some of my own paraphrasing / clarficiations in [brackets], because I did have some additional back and forth with him, and I don't want to claim that all these are his exact original words.

----begin quote------
...1 MB would probably service 1000 clients.  1GB would handle close to one million.
...
 I would [recommend making a overly conservative statement to admins] like "4GB should support several thousand concurrent clients" [in order to suggest] the basis of scaling and to leave lots of margin for error.

But maybe saying a bigger number is sensible....  With cheap disk space today, ...maybe "16GB should support several thousand concurrent clients" would be just as good, and much safer.
----end quote------

So I guess we get to make a final decision.  The doc was recently changed from 20 GB to 4GB, and that still sounds like plenty. But if we want to go 8 - 16GB that is justifiable too, since space is cheap and we don't know how many NFS shares (each with it's own NFS clients) will eventually be added to someone's cluster that started out smaller.

File: xml/article_nfs_storage.xml
Comment (tahliar) [reply xml/article_nfs_storage.xml:289]: Can we just leave out `resource-and-stonith` and `crm-fence-peer.sh`? What's the impact of not having them?

They're not included in the example `.res` files you attached in the bug, but I kept them from the original doc because they looked important :shrug: but that may have been a misjudgement on my part. 

I think I need to completely revisit the `.res` configs. What I did before was keep the configs from the original doc and change what seemed necessary based on your example configs. But now I realise I should have started by completely replacing the old config with your new example config, and then checked if anything from the old config should be added back.

**Comment by tahliar** [comment]
> I've heard back from Neil about the size of /var/lib/nfs. It's kind of a mixed message. I will quote him below but insert some of my own paraphrasing / clarficiations in [brackets], because I did have some additional back and forth with him, and I don't want to claim that all these are his exact original words.
> 
> ----begin quote------ ...1 MB would probably service 1000 clients. 1GB would handle close to one million. ... I would [recommend making a overly conservative statement to admins] like "4GB should support several thousand concurrent clients" [in order to suggest] the basis of scaling and to leave lots of margin for error.
> 
> But maybe saying a bigger number is sensible.... With cheap disk space today, ...maybe "16GB should support several thousand concurrent clients" would be just as good, and much safer. ----end quote------
> 
> So I guess we get to make a final decision. The doc was recently changed from 20 GB to 4GB, and that still sounds like plenty. But if we want to go 8 - 16GB that is justifiable too, since space is cheap and we don't know how many NFS shares (each with it's own NFS clients) will eventually be added to someone's cluster that started out smaller.

I tend to agree with Neil's last paragraph; better to err on the side of caution and use a bigger number for the example, along with a little explanation so people can decide for themselves if they want to use less. Happy to change this to whatever number we think is reasonable.

File: xml/article_nfs_storage.xml
Comment (zzhou1) [reply xml/article_nfs_storage.xml:289]: Now, this discussion start pulling in multiple advanced DRBD factors to be considered alone, 

> I'd like our steps to be scalable. With multiple files systems shared via NFS, is there a way to combine those all into one resource as well, for purposes of this concern?

1. Scalability is a good angle. Yes, DRBD is flexible to make it work to add more DRBD volumes with individual filesystem on each. However, a number of steps is needed with caution. Besides that, because of such flexibility, it can have variants for different purposes. For example, if "scalable" its self means the minimal interruption to the service to scale online, that will add further complexity to administration .

> Can we just leave out resource-and-stonith and crm-fence-peer.sh? What's the impact of not having them?

Good question. This is the more in-depth of DRBD internals. `man drbd.conf` tells more. Put it simple, resource-and-stonith is the fast approach to react on potential split-brain in the DRBD cluster. But it comes with limitation just like we are facing. 

2. Simplicity is another angle I think this DOC trying to be. With that in mind, yes, have "multiple DRBD resources without resource-and-stonith" need much less steps to scale to multiple filesystems. 

3. Resilience level can be different for DRBD Clusters . Besides the "resource-and-stonith" approach, also there are other approaches,   eg. "automatic split brain policy" in my example attached in the bug. Even, some customers equip both.

Circle back to the motivation of this effort to improve NFS DRBD use case to be more resilient on failures. I slightly incline to setup "resource-and-stonith".  

In the next few days, I could do more homework and come up with the appropriate minimum steps about how to add DRBD volumes for this NFS use case. It could show us its complexity level, and judge if worth to DOC it. Or, lucky enough to find a better way out.

Anyway, it is still open for more comments, more discussion, and debate.

Comment (zzhou1) [reply xml/article_nfs_storage.xml:289]: Here below is my example steps to scale one more filesystem into the running cluster. That sounds doable for this DOC.

"""

    1. crm cluster run "lvcreate -n share2 -L 1G nfs"

    2. Update nfs.res to add new volume. Caution on: vol-number, minor-X, /dev/nfs/shareX
      >     volume 2 {
      >         device           /dev/drbd3;
      >         disk             /dev/nfs/share2;

      cysnc2 the configuration.

    3. crm cluster run "drbdadm create-md nfs/2 --force"

    4. crm cluster run "drbdadm adjust nfs"

    5. drbdadm new-current-uuid --clear-bitmap nfs/2

    6. mkfs.ext4 /dev/drbd3

    7. crm config edit to add new filesystem and exportfs etc.

"""

Here are the reference CIB and drbd .res files: 
[nfs-1more-share.0322.cib.txt](https://github.com/SUSE/doc-sleha/files/11041218/nfs-1more-share.0322.cib.txt)
[nfs-1more-share.0322.res.txt](https://github.com/SUSE/doc-sleha/files/11041220/nfs-1more-share.0322.res.txt)

NOTE:  

a) drbd nfs.res consolidate all into multi-volume drbd devices. Hence <resources/vol-number> designation is used often now. 

b)  For the cluster configuration CIB, I stretch my NFS knowledge again, and fail the attempt to isolate the failure of one filesystem to leave other nfs shares keep providing service. Basically, I restore CIB to the plain group resource type. That says, stopping one filesystem will impact the whole cluster, for example.

Comment (tahliar) [reply xml/article_nfs_storage.xml:289]: Thanks for those steps, Roger. I think that's the ideal way to present it: as a final section showing an example of how to add more filesystems to the config. 

Regarding groups:

From what I understand, resources in a group are started in the order they're added to the group (and stopped in reverse order). This is convenient for addressing point 3 in the TID, where the resources need to be started in a specific order. The group setup is also much simpler to present in the doc, so I'm on board with using a group.

~However, when we create the new filesystem and exportfs, does that mean that adding them to the group last means they'll be started last? How do we adjust the group so that the new filesystem and exportfs slot in to the appropriate starting order?~ 

Actually, `crm help configure modgroup` had the answer. We can use `before` or `after` to specify where the new resources go in the group. Solved :+1:

**Comment by tahliar** [comment]
@zzhou1 @dpartrid I've updated the steps to use a single DRBD config file, based on Roger's example steps. I've also added a section for adding new NFS shares to the cluster. Once you're both happy with these steps, then I'll move it on to Neil to review as well. 

I tested this setup in my test cluster, and can confirm the steps work in the sense that I successfully created the resources and group. I can't confirm whether they *work* in the broader sense of solving the issues from the TID, but the concept seems sound to me. 

I did run into an issue where `p_nfsserver` would only start on the first node, but failed to start if the group moved to the second node (even though `nfs-kernel-server` was installed on both nodes). But this could be a problem with my test cluster, as I've used it for other tests previously.
In any case, once everyone approves of the steps in the doc, I think it would be a good idea to have it go through some thorough QA to check for any other potential issues before publishing.

File: xml/article_nfs_storage.xml
Comment (tahliar) [inline xml/article_nfs_storage.xml:214]: Could these be `/dev/drbd0` and `/dev/drbd1`? I just wonder if having mismatched numbers for the volumes and devices might potentially trip anyone up. Could be easy to mistype the wrong number.

File: xml/article_nfs_storage.xml
Comment (tahliar) [inline xml/article_nfs_storage.xml:502]: Is all of this required for this specific setup, or is some of it incidental? (Same question for the other resources, too).

File: xml/article_nfs_storage.xml
Comment (zzhou1) [reply xml/article_nfs_storage.xml:214]: Aha, exactly. Coincidentally, I was reviewing another bug and correlated to this about an hour ago.  

Yes, your suggestion is better. For example,

volume 0 <--> drbd0 <--> /nfs/state
volume 1 <--> drbd1 <--> /nfs/share

File: xml/article_nfs_storage.xml
Comment (zzhou1) [reply xml/article_nfs_storage.xml:502]: IIRC, they are auto-populated by crmsh, based on advertisement of each specific RA.

File: xml/article_nfs_storage.xml
Comment (dpartrid) [reply xml/article_nfs_storage.xml:214]: Just FYI:  Besides section 4.1, section 4.2 mentions drbd1/2 as well.

**Comment by dpartrid** [comment]
> 
I'm not familiar enough with this to comment well.  I would have thought that as long as /var/lib/nfs is present BEFORE nfsserver and exportfs are activated, all the necessary parts within it would get created as needed.  If it looks like I'm misunderstanding the concern here, please give me the "restated for dummies" version.  :-)

**Comment by dpartrid** [comment]
early in section 5, this is mentioned:

...an NFS server which serves clients in the 192.168.1.x/24 subnet

but when the exportfs primitive is later created, the clientspec given is "*" which means all clients, not just clients in the one subnet.  To fit the previous statement, the client spec should be "192.168.1.0/24"

File: xml/article_nfs_storage.xml
Comment (tahliar) [reply xml/article_nfs_storage.xml:214]: I found conflicting info about whether `minor x` can start at 0 or 1, but found consistent info that if you specify a name then you can leave out the minor. So I'm removing `minor 1` and `minor 2`, but do let me know if it actually needs to stay in.

**Comment by tahliar** [comment]
Updated again with small fixes from the discussions above. I've added `nfs_shared_infodir="/var/lib/nfs"` to the `nfsserver` steps; maybe later we can remove it when https://github.com/ClusterLabs/resource-agents/pull/1860 is resolved.

**Comment by tahliar** [comment]
@dariavladykina This is an extensive overhaul, so I've kept the PDF up to date in the description; hopefully it helps. 

Soft deadline for this PR is the June 6 priority 3 proofing drop; hard deadline is the as-yet-TBD priority 3 localisation drop.

**Review by dariavladykina — COMMENTED** [review]
Hi, please see some minor suggestions (mainly commas). Thanks!

File: xml/article_nfs_storage.xml
Comment (tahliar) [reply xml/article_nfs_storage.xml:706]: Not picky at all! This was definitely a typo, not an intentional choice. Thanks for catching it!

**Pull request merged**
